File: lib/models/watch_room_model.dart
class WatchRoom {
  final String id;
  final String roomCode;
  final String movieSlug;
  final String movieName;
  final String moviePoster;
  final String hostId;
  final String hostName;
  final List<Participant> participants;
  final int maxParticipants;
  final bool isActive;
  final double currentTime;
  final bool isPlaying;
  final int currentServer;
  final int currentEpisode;
  final DateTime createdAt;

  WatchRoom({
    required this.id,
    required this.roomCode,
    required this.movieSlug,
    required this.movieName,
    required this.moviePoster,
    required this.hostId,
    required this.hostName,
    required this.participants,
    required this.maxParticipants,
    required this.isActive,
    required this.currentTime,
    required this.isPlaying,
    required this.currentServer,
    required this.currentEpisode,
    required this.createdAt,
  });

  factory WatchRoom.fromJson(Map<String, dynamic> json) {
    return WatchRoom(
      id: json['_id'] ?? '',
      roomCode: json['roomCode'] ?? '',
      movieSlug: json['movieSlug'] ?? '',
      movieName: json['movieName'] ?? '',
      moviePoster: json['moviePoster'] ?? '',
      hostId: json['host'] ?? '',
      hostName: json['hostName'] ?? '',
      participants:
          (json['participants'] as List<dynamic>?)
              ?.map((p) => Participant.fromJson(p))
              .toList() ??
          [],
      maxParticipants: json['maxParticipants'] ?? 10,
      isActive: json['isActive'] ?? true,
      currentTime: (json['currentTime'] ?? 0).toDouble(),
      isPlaying: json['isPlaying'] ?? false,
      currentServer: json['currentServer'] ?? 0,
      currentEpisode: json['currentEpisode'] ?? 0,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'roomCode': roomCode,
      'movieSlug': movieSlug,
      'movieName': movieName,
      'moviePoster': moviePoster,
      'host': hostId,
      'hostName': hostName,
      'participants': participants.map((p) => p.toJson()).toList(),
      'maxParticipants': maxParticipants,
      'isActive': isActive,
      'currentTime': currentTime,
      'isPlaying': isPlaying,
      'currentServer': currentServer,
      'currentEpisode': currentEpisode,
      'createdAt': createdAt.toIso8601String(),
    };
  }

  int get participantCount => participants.length;
  bool get isFull => participants.length >= maxParticipants;
}

class Participant {
  final String id;
  final String name;
  final String? avatar;
  final DateTime joinedAt;

  Participant({
    required this.id,
    required this.name,
    this.avatar,
    required this.joinedAt,
  });

  factory Participant.fromJson(Map<String, dynamic> json) {
    return Participant(
      id: json['user'] ?? json['_id'] ?? '',
      name: json['name'] ?? '',
      avatar: json['avatar'],
      joinedAt: json['joinedAt'] != null
          ? DateTime.parse(json['joinedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user': id,
      'name': name,
      'avatar': avatar,
      'joinedAt': joinedAt.toIso8601String(),
    };
  }
}

// Video sync state model
class VideoSyncState {
  final double currentTime;
  final bool isPlaying;
  final int currentServer;
  final int currentEpisode;
  final String? triggeredBy;

  VideoSyncState({
    required this.currentTime,
    required this.isPlaying,
    required this.currentServer,
    required this.currentEpisode,
    this.triggeredBy,
  });

  factory VideoSyncState.fromJson(Map<String, dynamic> json) {
    return VideoSyncState(
      currentTime: (json['currentTime'] ?? 0).toDouble(),
      isPlaying: json['isPlaying'] ?? false,
      currentServer: json['currentServer'] ?? 0,
      currentEpisode: json['currentEpisode'] ?? 0,
      triggeredBy: json['triggeredBy'],
    );
  }
}

File: server/src/models/WatchRoom.model.js
import mongoose from 'mongoose';

const participantSchema = new mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    name: {
        type: String,
        required: true
    },
    avatar: {
        type: String,
        default: null
    },
    joinedAt: {
        type: Date,
        default: Date.now
    }
});

const watchRoomSchema = new mongoose.Schema({
    roomCode: {
        type: String,
        required: true,
        unique: true,
        minlength: 6,
        maxlength: 6
    },
    movieSlug: {
        type: String,
        required: true
    },
    movieName: {
        type: String,
        required: true
    },
    moviePoster: {
        type: String,
        default: ''
    },
    host: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    hostName: {
        type: String,
        required: true
    },
    participants: [participantSchema],
    maxParticipants: {
        type: Number,
        default: 30,
        min: 2,
        max: 50
    },
    isActive: {
        type: Boolean,
        default: true
    },
    // Video sync state
    currentTime: {
        type: Number,
        default: 0
    },
    isPlaying: {
        type: Boolean,
        default: false
    },
    currentServer: {
        type: Number,
        default: 0
    },
    currentEpisode: {
        type: Number,
        default: 0
    }
}, {
    timestamps: true
});

// Index for faster queries
watchRoomSchema.index({ roomCode: 1 });
watchRoomSchema.index({ isActive: 1 });
watchRoomSchema.index({ host: 1 });

// Generate unique room code
watchRoomSchema.statics.generateRoomCode = async function () {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    let code;
    let exists = true;

    while (exists) {
        code = '';
        for (let i = 0; i < 6; i++) {
            code += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        exists = await this.findOne({ roomCode: code, isActive: true });
    }

    return code;
};

const WatchRoom = mongoose.model('WatchRoom', watchRoomSchema);

// Drop stale roomId index if exists (fix for duplicate key error)
WatchRoom.collection.dropIndex('roomId_1').catch(() => {
    // Index doesn't exist, ignore error
});

export default WatchRoom;
